### PR TITLE
Add Pokédex extension

### PR DESCRIPTION
**Pull request by anhthang**
## Description

Search for a Pokémon by name or using its National Pokédex number.

## Type of change

- New extension

## Screencast

https://user-images.githubusercontent.com/10458061/153992584-d7680b05-6a54-46e9-b550-86845b36f16d.mov

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

### CODE REVIEW

File: extensions/pokedex/src/api/index.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
+import axios, { AxiosRequestConfig } from "axios";
+import { showToast, ToastStyle } from "@raycast/api";
+import type { PokemonV2Pokemon } from "../types";
+
+export const getPokemon = async (
+  nameOrId: string
+): Promise<PokemonV2Pokemon[]> => {
+  const condition = Number.isNaN(Number(nameOrId))
+    ? `{name: {_eq: ${nameOrId}}}`
```
[the remark this answers is not in the document]
Comment (anhthang) [reply extensions/pokedex/src/api/index.ts:9]: Thanks for letting me know about this. I have fixed it in latest commit.

> Can't search for “Pikachu”, only “pikachu”

**Comment by anhthang** [comment]
Hi @pernielsentikaer

Thank you for your suggestions. I have moved to `media` folder. Please help me review again.

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @anhthang 👋 

You added the .toLowerCase() to `id`, not the `name`

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by anhthang** [comment]
Hi @pernielsentikaer 

I have fixed. It's my mistaken when sync from my repository to raycast/extensions

**Review by pernielsentikaer — COMMENTED** [review]
Hi @anhthang 👋 

Thanks for the correction 💯 After testing it out, I keep getting `not a valid graphql query` both when searching for number or name – even if I revert the latest push 

Do you know if anything changed in the other end?

**Comment by anhthang** [comment]
Hi @pernielsentikaer 

Thanks for letting me know. I will do more test again to figure out what is error with graphql query as I haven't get this issue before.

**Comment by anhthang** [comment]
Hi @pernielsentikaer 

Yes, it will be really nice. I also have some to do items to improve the extension that I will do later.

Thanks for your help.

**Comment by anhthang** [comment]
Hi @pernielsentikaer 

I have done your suggestion and updated. Could you please review again? 😊

Also, is there any way to disable search box in the List?

Screen Shot 2022-02-12 at 12 09 16 AM

**Comment by pernielsentikaer** [comment]
Hi @anhthang 👋 

It works much better now, thanks for the update 🔥 

The only way to avoid the search in top is to use a [Detail view with Markdown](https://developers.raycast.com/api-reference/user-interface/detail#detail) - something like this:

https://user-images.githubusercontent.com/38033107/153670163-2b1e0efb-1a0f-49bf-b9b7-f9629bdbed71.mp4

I'm not sure the list with attributes looks as smooth as when it's a list, let me know what you think - it's kinda also awesome as detail with the image

Right now, when you change from one list to the next, you have [empty state flicker](https://developers.raycast.com/basics/prepare-an-extension-for-store#empty-states) that we need to address

**Comment by anhthang** [comment]
Hi @pernielsentikaer 👋 

Yeah, It's much better when we use markdown to render and show image. So I changed it. But it seems CommonMark is not support table, that would be better a lots.

image

Loading state added already but I'm not sure that the flicker still there. Please help me double check on it.

**Comment by pernielsentikaer** [comment]
That turned out quite nice @anhthang 💪 

In your surprise me, I would push the PokemonDetail directly, do you think you can work that out? That would improve the UX a lot 🚀

**Comment by anhthang** [comment]
Hi @pernielsentikaer 

Thanks for the idea, I forgot that we can do it 😄 The extension updated just now.

File: extensions/pokedex/package.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    {
+      "name": "index",
+      "title": "Search Pokémon",
+      "subtitle": "Pokédex",
+      "description": "Search for a Pokémon by name or using its National Pokédex number.",
+      "mode": "view"
+    }
+  ],
+  "dependencies": {
+    "@raycast/api": "^1.27.1",
```
Comment (pernielsentikaer) [inline extensions/pokedex/package.json:19]:
```suggestion
    "@raycast/api": "^1.28.0",
```

File: extensions/pokedex/src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            subtitle={pokemons.length.toString()}
+          >
+            {pokemons.map((pokemon) => (
+              <List.Item
+                key={pokemon.id}
+                title={`#${pokemon.id.toString().padStart(3, "0")}`}
+                subtitle={pokemon.name}
+                accessoryTitle={pokemon.types.join(", ")}
+                icon={{
+                  source: pokemon.artwork,
```
Comment (pernielsentikaer) [inline extensions/pokedex/src/index.tsx:91]:
```suggestion
                  source: pokemon.artwork,
                  fallback: "icon.png"
```
Fallback if image fails to load

File: extensions/pokedex/src/index.tsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                }}
+                actions={
+                  <ActionPanel>
+                    <ActionPanel.Item
+                      title="Show Details"
```
Comment (pernielsentikaer) [inline extensions/pokedex/src/index.tsx:96]:
```suggestion
                      title="Show Details",
                      icon={Icon.MagnifyingGlass},
```
Or another icon

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @anhthang 👋 

I think we're about being there, It's turning out quite nice – what a UX!

Check the latest comments, could be an idea to add fallback image just in case :)

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @anhthang 👋 

After last review, I'm not able to load almost any Detail views, somehow only 001 opens, the rest keeps spinning forever (the same with surprise me) – can you check if anything is wrong?

When setting a `console.log` in `PokemonDetail` – I can see that Bulbasaur logs 4 times, rest only two, I'm not sure it has anything to do with it?

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by anhthang** [comment]
Hi @pernielsentikaer 👋 

I have tested again and got different Pokemon every times. I'm not sure why it happens to you, maybe it's just a coincidence.

Regarding to the loading, I think the API is unstable at that time and keeping the connection. IMO, we can set a timeout and show no results view for this in the next PR for this case.

**Comment by pernielsentikaer** [comment]
Hi @anhthang 👋 

I think you're right! Everything seems to work as expected right now 💪

Do you have an idea to show a failure toast when it takes too long time to get the response, it sounds like an awesome idea 🚀

**Comment by anhthang** [comment]
Hi @pernielsentikaer 

I think it just a toast with message like: "Unable to get Pokémon details." for all the error from PokéAPI and a page with no results image. Do we have a thing like that for Detail API?

And, how long do you think it's good to wait? Is it 10s is enough as it usually getting result within 1-2s.

**Comment by pernielsentikaer** [comment]
If it's possible, I think it should time out after around 5 seconds, as you says it shouldn't take that long time.

The message could be something with “…try again later”

**Comment by anhthang** [comment]
Hi @pernielsentikaer 

What do you think about showing a failure toast, then back to the listing page if error?

**Review by pernielsentikaer — APPROVED** [review]
Hi @anhthang 🔥 

Awesome!!! Wow, it turned out even more awesome that I expected, well done 💪

**Pull request merged**